Stochastic pmap lowering behavior in tests

### Description

I'm working on debugging a failure that we're seeing in our daily open source build. [This test](https://github.com/google/jax/blob/65063b7ac30635e76d2d692fb976a11f136e32bf/tests/pmap_test.py#L2068) is consistently failing at `main` when we run the full test suite with 4 GPUs using `pytest` with the error:

```
tests/pmap_test.py:2084: in test_grad_of_pmap_compilation_caching
    self.assertEqual(count[0], 2)  # one for fwd, one for bwd
E   AssertionError: 3 != 2
```

This failure can't be reproduced with a single test and it only appeared last week, although I think the underlying issue must be older than that. I thought it might be some shared state, but I've isolated the issue a little bit. When logging the compiles, I find that the three lowering operations are:

1. [`pmap_f`](https://github.com/google/jax/blob/65063b7ac30635e76d2d692fb976a11f136e32bf/tests/pmap_test.py#L2082),
2. [`pmap_backwards_pass`](https://github.com/google/jax/blob/65063b7ac30635e76d2d692fb976a11f136e32bf/tests/pmap_test.py#L2083) (transpose of `f`), and
3. `jit__multi_slice` (defined [here](https://github.com/google/jax/blob/85f6ba1693e1e547335f3de93230b3d956dbae70/jax/_src/numpy/array_methods.py#L318)).

When I run just this test, I find that `jit__multi_slice` should be lowered during the "warm-up" call here:

https://github.com/google/jax/blob/65063b7ac30635e76d2d692fb976a11f136e32bf/tests/pmap_test.py#L2079

but, when the test fails, it's not actually being compiled until here:

https://github.com/google/jax/blob/65063b7ac30635e76d2d692fb976a11f136e32bf/tests/pmap_test.py#L2083

It seems like maybe there's some kind of async race condition or something here, but adding a `block_until_ready` didn't fix the issue. I wanted to raise this here to see if anyone had thoughts for where this might be coming from, but I know that it's annoying to repro, so no worries if not!

cc @mattjj specifically as the author of this test

### System info (python version, jaxlib version, accelerator, etc.)

TBD

--- a/tests/pmap_test.py
+++ b/tests/pmap_test.py
@@ -2075,13 +2075,10 @@ class PythonPmapTest(jtu.JaxTestCase):
     def f(x):
       return jnp.sin(x)
 
+    # warm-up the cache
     x = jnp.ones(axis_size)
-    f(x)  # warm-up any dispatching compilations
-
-    with jtu.count_jit_and_pmap_compiles() as count:  # noqa: F841
-      _, f_bwd  = jax.vjp(f, x)
-      _ = f_bwd(x)
-    self.assertEqual(count[0], 2)  # one for fwd, one for bwd
+    _, f_bwd  = jax.vjp(f, x)
+    _ = f_bwd(x)
 
     with jtu.count_jit_and_pmap_compiles() as count:  # noqa: F841
       _, f_bwd2  = jax.vjp(f, x)

